From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Tue, 5 Aug 2025 02:26:29 +0900
Subject: chore: restore some deprecated wrapper utility in gin

Restores part of https://chromium-review.googlesource.com/c/chromium/src/+/6799157

Patch can be removed once cppgc migration is complete
https://github.com/electron/electron/issues/47922

diff --git a/gin/function_template.h b/gin/function_template.h
index 84ab9585240a49048774811718f7ebd6f988e485..f062163cdd81def12fae7e507d18a9133dd0804d 100644
--- a/gin/function_template.h
+++ b/gin/function_template.h
@@ -77,6 +77,7 @@ class GIN_EXPORT CallbackHolderBase {
                     CallbackHolderBase* holder);
     ~DisposeObserver() override;
     void OnBeforeDispose(v8::Isolate* isolate) override;
+    void OnBeforeMicrotasksRunnerDispose(v8::Isolate* isolate) override {}
     void OnDisposed() override;
 
    private:
diff --git a/gin/isolate_holder.cc b/gin/isolate_holder.cc
index bb1639d73070a99984b72eb61afd001dec5b08ff..b036f324309c46c53b74124da3ea830d39a973e3 100644
--- a/gin/isolate_holder.cc
+++ b/gin/isolate_holder.cc
@@ -224,6 +224,7 @@ void IsolateHolder::WillCreateMicrotasksRunner() {
 
 void IsolateHolder::WillDestroyMicrotasksRunner() {
   DCHECK(g_initialized_microtasks_runner);
+  isolate_data_->NotifyBeforeMicrotasksRunnerDispose();
   g_destroyed_microtasks_runner = true;
 }
 
diff --git a/gin/object_template_builder.cc b/gin/object_template_builder.cc
index 5a31687bbd0fca61db3a7c41ed73d938340d6446..b84f5fd336bc4b61b2cd0b2fc92382b00e928701 100644
--- a/gin/object_template_builder.cc
+++ b/gin/object_template_builder.cc
@@ -141,7 +141,7 @@ ObjectTemplateBuilder::ObjectTemplateBuilder(v8::Isolate* isolate,
                   : "Objects of this type cannot be created using the "
                     "constructor"))),
       template_(constructor_template_->InstanceTemplate()) {
-  template_->SetInternalFieldCount(0);
+  template_->SetInternalFieldCount(kNumberOfInternalFields);
 }
 
 ObjectTemplateBuilder::ObjectTemplateBuilder(v8::Isolate* isolate,
diff --git a/gin/per_isolate_data.cc b/gin/per_isolate_data.cc
index 884990426f13a6abca22a60dd8cc0685f8435b23..d1014af4b63da244820ff865a8e824ddf68433a9 100644
--- a/gin/per_isolate_data.cc
+++ b/gin/per_isolate_data.cc
@@ -68,12 +68,37 @@ PerIsolateData* PerIsolateData::From(Isolate* isolate) {
   return static_cast<PerIsolateData*>(isolate->GetData(kEmbedderNativeGin));
 }
 
+void PerIsolateData::DeprecatedSetObjectTemplate(DeprecatedWrapperInfo* info,
+                                                 Local<ObjectTemplate> templ) {
+  deprecated_object_templates_[info] = Eternal<ObjectTemplate>(isolate_, templ);
+}
+
 void PerIsolateData::SetObjectTemplate(
     const WrapperInfo* info,
     Local<ObjectTemplate> templ) {
   object_templates_[info] = Eternal<ObjectTemplate>(isolate_, templ);
 }
 
+void PerIsolateData::DeprecatedSetFunctionTemplate(
+    DeprecatedWrapperInfo* info, Local<FunctionTemplate> templ) {
+  deprecated_function_templates_[info] = Eternal<FunctionTemplate>(isolate_, templ);
+}
+
+void PerIsolateData::SetFunctionTemplate(
+    const WrapperInfo* info, Local<FunctionTemplate> templ) {
+  function_templates_[info] = Eternal<FunctionTemplate>(isolate_, templ);
+}
+
+v8::Local<v8::ObjectTemplate> PerIsolateData::DeprecatedGetObjectTemplate(
+    DeprecatedWrapperInfo* info) {
+  DeprecatedObjectTemplateMap::iterator it =
+      deprecated_object_templates_.find(info);
+  if (it == deprecated_object_templates_.end()) {
+    return v8::Local<v8::ObjectTemplate>();
+  }
+  return it->second.Get(isolate_);
+}
+
 v8::Local<v8::ObjectTemplate> PerIsolateData::GetObjectTemplate(
     const WrapperInfo* info) {
   ObjectTemplateMap::iterator it = object_templates_.find(info);
@@ -83,6 +108,25 @@ v8::Local<v8::ObjectTemplate> PerIsolateData::GetObjectTemplate(
   return it->second.Get(isolate_);
 }
 
+v8::Local<v8::FunctionTemplate> PerIsolateData::DeprecatedGetFunctionTemplate(
+    DeprecatedWrapperInfo* info) {
+  DeprecatedFunctionTemplateMap::iterator it =
+      deprecated_function_templates_.find(info);
+  if (it == deprecated_function_templates_.end()) {
+    return v8::Local<v8::FunctionTemplate>();
+  }
+  return it->second.Get(isolate_);
+}
+
+v8::Local<v8::FunctionTemplate> PerIsolateData::GetFunctionTemplate(
+    const WrapperInfo* info) {
+  FunctionTemplateMap::iterator it = function_templates_.find(info);
+  if (it == function_templates_.end()) {
+    return v8::Local<v8::FunctionTemplate>();
+  }
+  return it->second.Get(isolate_);
+}
+
 void PerIsolateData::AddDisposeObserver(DisposeObserver* observer) {
   dispose_observers_.AddObserver(observer);
 }
@@ -97,6 +141,12 @@ void PerIsolateData::NotifyBeforeDispose() {
   }
 }
 
+void PerIsolateData::NotifyBeforeMicrotasksRunnerDispose() {
+  for (auto& observer : dispose_observers_) {
+    observer.OnBeforeMicrotasksRunnerDispose(isolate_.get());
+  }
+}
+
 void PerIsolateData::NotifyDisposed() {
   for (auto& observer : dispose_observers_) {
     observer.OnDisposed();
diff --git a/gin/per_isolate_data.h b/gin/per_isolate_data.h
index bce889749415da341e6e6e4082ac06bbeb4bb80a..2f8abc344c77713fb10d83e51ba486c84ab93474 100644
--- a/gin/per_isolate_data.h
+++ b/gin/per_isolate_data.h
@@ -34,6 +34,10 @@ class GIN_EXPORT PerIsolateData {
     // be entered before the observer is notified, but there will not be a
     // handle scope by default.
     virtual void OnBeforeDispose(v8::Isolate* isolate) = 0;
+
+    // Called just before the microtasks runner is about to be disposed.
+    virtual void OnBeforeMicrotasksRunnerDispose(v8::Isolate* isolate) = 0;
+
     // Called just after the isolate has been disposed.
     virtual void OnDisposed() = 0;
   };
@@ -51,14 +55,36 @@ class GIN_EXPORT PerIsolateData {
 
   static PerIsolateData* From(v8::Isolate* isolate);
 
+  void DeprecatedSetObjectTemplate(
+      DeprecatedWrapperInfo* info,
+      v8::Local<v8::ObjectTemplate> object_template);
+
   void SetObjectTemplate(const WrapperInfo* info,
                          v8::Local<v8::ObjectTemplate> object_template);
 
+  void DeprecatedSetFunctionTemplate(
+      DeprecatedWrapperInfo* info,
+      v8::Local<v8::FunctionTemplate> function_template);
+
+  void SetFunctionTemplate(
+      const WrapperInfo* info,
+      v8::Local<v8::FunctionTemplate> function_template);
+
+  v8::Local<v8::ObjectTemplate> DeprecatedGetObjectTemplate(
+      DeprecatedWrapperInfo* info);
+
   v8::Local<v8::ObjectTemplate> GetObjectTemplate(const WrapperInfo* info);
 
+  v8::Local<v8::FunctionTemplate> DeprecatedGetFunctionTemplate(
+      DeprecatedWrapperInfo* info);
+
+  v8::Local<v8::FunctionTemplate> GetFunctionTemplate(
+      const WrapperInfo* info);
+
   void AddDisposeObserver(DisposeObserver* observer);
   void RemoveDisposeObserver(DisposeObserver* observer);
   void NotifyBeforeDispose();
+  void NotifyBeforeMicrotasksRunnerDispose();
   void NotifyDisposed();
 
   void EnableIdleTasks(std::unique_ptr<V8IdleTaskRunner> idle_task_runner);
@@ -74,14 +100,23 @@ class GIN_EXPORT PerIsolateData {
   }
 
  private:
+  typedef std::map<DeprecatedWrapperInfo*, v8::Eternal<v8::ObjectTemplate>>
+      DeprecatedObjectTemplateMap;
   typedef std::map<const WrapperInfo*, v8::Eternal<v8::ObjectTemplate>>
       ObjectTemplateMap;
+  typedef std::map<DeprecatedWrapperInfo*, v8::Eternal<v8::FunctionTemplate>>
+      DeprecatedFunctionTemplateMap;
+  typedef std::map<const WrapperInfo*, v8::Eternal<v8::FunctionTemplate>>
+      FunctionTemplateMap;
 
   // PerIsolateData doesn't actually own |isolate_|. Instead, the isolate is
   // owned by the IsolateHolder, which also owns the PerIsolateData.
   raw_ptr<v8::Isolate, AcrossTasksDanglingUntriaged> isolate_;
   raw_ptr<v8::ArrayBuffer::Allocator, DanglingUntriaged> allocator_;
+  DeprecatedObjectTemplateMap deprecated_object_templates_;
   ObjectTemplateMap object_templates_;
+  DeprecatedFunctionTemplateMap deprecated_function_templates_;
+  FunctionTemplateMap function_templates_;
   base::ObserverList<DisposeObserver> dispose_observers_;
   std::shared_ptr<V8ForegroundTaskRunnerBase> task_runner_;
   std::shared_ptr<V8ForegroundTaskRunnerBase> user_visible_task_runner_;
diff --git a/gin/public/wrapper_info.h b/gin/public/wrapper_info.h
index 34b5f1c30c05152122f23708a1df62f00296fcd6..dac3459dc822db1b242288293605ab4c4a6cf76f 100644
--- a/gin/public/wrapper_info.h
+++ b/gin/public/wrapper_info.h
@@ -13,6 +13,17 @@
 
 namespace gin {
 
+enum InternalFields {
+  kWrapperInfoIndex,
+  kEncodedValueIndex,
+  kNumberOfInternalFields,
+};
+
+struct GIN_EXPORT DeprecatedWrapperInfo {
+  static DeprecatedWrapperInfo* From(v8::Local<v8::Object> object);
+  const GinEmbedder embedder;
+};
+
 struct GIN_EXPORT WrapperInfo : v8::Object::WrapperTypeInfo {
   const WrappablePointerTag pointer_tag;
 };
diff --git a/gin/wrappable.cc b/gin/wrappable.cc
index 803b5648e1c3ec3621149e98850bebfbf7f2de75..dd49202103993ee03379acd6873b92bc8fccb786 100644
--- a/gin/wrappable.cc
+++ b/gin/wrappable.cc
@@ -48,7 +48,7 @@ v8::MaybeLocal<v8::Object> WrappableBase::GetWrapper(v8::Isolate* isolate) {
     CHECK(!templ.IsEmpty());
     data->SetObjectTemplate(info, templ);
   }
-  CHECK_EQ(0, templ->InternalFieldCount());
+  CHECK_EQ(kNumberOfInternalFields, templ->InternalFieldCount());
   v8::Local<v8::Object> wrapper;
   // |wrapper| may be empty in some extreme cases, e.g., when
   // Object.prototype.constructor is overwritten.
@@ -56,6 +56,12 @@ v8::MaybeLocal<v8::Object> WrappableBase::GetWrapper(v8::Isolate* isolate) {
     return {};
   }
 
+  // Delete the internal fields once gin_helper::DeprecatedWrappable does
+  // not exist anymore.
+  int indices[] = {kWrapperInfoIndex, kEncodedValueIndex};
+  void* values[] = {nullptr, nullptr};
+  wrapper->SetAlignedPointerInInternalFields(2, indices, values);
+
   AssociateWithWrapper(isolate, wrapper);
   return wrapper;
 }
